### ISSUE TITLE
Perl5.8.0+ Data::Dumper when using $Sortkeys causes subsequent each() failures

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8236]
Migrated from [rt.perl.org#37854](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=37854) (status was 'resolved')

Searchable as RT37854$

**Comment by p5pRT** [comment]
### From jerry.hebert@gmail.com

#### Created by jerry.hebert@gmail.com


This is a bug report for perl from jerry\.hebert@&#8203;gmail\.com\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.6\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
Hello\, I have discovered an error in Data&#8203;::Dumper on versions
greater than 5\.8\.0 Perl\. After 5\.8\.0\, Data&#8203;::Dumper uses qsortsv\(\)
to sort keys if $Sortkeys is true and prior to that it used a
pure Perl sort\. If you are dumping a hash\, this somehow causes
problems with the hash iterator when attempting to subsequently
use each\(\)\.

The following is a small script to illustrate the behavior\.

\#\!/usr/local/bin/perl
use strict;
use warnings;

use Data&#8203;::Dumper;

my $opts = \(  x => 5 \);

$Data&#8203;::Dumper&#8203;::Sortkeys = 1;
my $dumped = Data&#8203;::Dumper\->Dump \(\[ \\%opts \]\);

\# Uncomment the following keys\(\) line to print each properly
\#keys %opts;
while \(my \($key\, $val\) = each %opts\) \{
&nbsp;    print "$key&#8203;: $val\\n";
\}

\_\_END\_\_

You will see that the each\(\) loop prints nothing\. If you change 
$Data&#8203;::Dumper&#8203;::Sortkeys to 0 or don't set it\, the 
behavior will go away\. Also\, using $Data&#8203;::Dumper&#8203;::Useperl =1
fixes the problem as well\.

Please let me know if you have further questions about this\.

thanks\,
jerry

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.6:

Configured by ghsu at Wed Aug 31 12:40:10 PDT 2005.

Summary of my perl5 (revision 5 version 8 subversion 6) configuration:
  Platform:
    osname=linux, osvers=2.4.21-27.0.1.elsmp, archname=opt-gccWS3
    uname='linux walrus 2.4.21-27.0.1.elsmp #1 smp mon dec 20 18:47:45 est 2004 i686 i686 i386 gnulinux '
    config_args='-des -Dlibs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lpthread -Dccflags=-fno-strict-aliasing -pipe -I/usr/local/include -pthread -Dinc_version_list=none -Dcf_email=pe@pdi.com -Duselargefiles -Dcc=/usr/bin/gcc -Dprefix=/rel/lang/perl/5.8.6 -Darchname=opt-gccWS3 -Dbin=/rel/lang/perl/5.8.6/opt-gccWS3/bin -Dscriptdir=/rel/lang/perl/5.8.6/opt-gccWS3/bin -Dprivlib=/rel/lang/perl/5.8.6/lib -Darchlib=/rel/lang/perl/5.8.6/opt-gccWS3/lib -Dman1dir=/rel/lang/perl/5.8.6/man/man1 -Dman3dir=/rel/lang/perl/5.8.6/man/man3 -Dsiteprefix=/rel/lang/perl/5.8.6 -Dsitebin=/rel/lang/perl/5.8.6/opt-gccWS3/bin -Dsitescript=/rel/lang/perl/5.8.6/opt-gccWS3/bin -Dsitelib=/rel/lang/perl/5.8.6/lib/site_perl -Dsitearch=/rel/lang/perl/5.8.6/opt-gccWS3/lib/site_perl -Dsiteman1dir=/rel/lang/perl/5.8.6/man/man1 -Dsiteman3dir=/rel/lang/perl/5.8.6/man/man3 -Doptimize=-O2'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='/usr/bin/gcc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include -pthread -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include -pthread -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.3 20030502 (Red Hat Linux 3.2.3-49)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='/usr/bin/gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lpthread
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc -lpthread
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.6:
    /rel/lang/perl/5.8.6/opt-gccWS3/lib
    /rel/lang/perl/5.8.6/lib
    /rel/lang/perl/5.8.6/opt-gccWS3/lib/site_perl
    /rel/lang/perl/5.8.6/lib/site_perl
    /rel/lang/perl/5.8.6/lib/site_perl
    .


Environment for perl v5.8.6:
    HOME=/home/jhebert
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LANGVAR=en_US.UTF-8
    LD_LIBRARY_PATH=/home/jhebert/lib.linux:/home/jhebert/lib:/home/oracle/product/8.1.7/lib
    LOGDIR (unset)
    PATH=.:/home/jhebert/bin:/home/jhebert/work/lib.src:/home/jhebert/work/cmd.src:/studio/shrek3/shdw_td/lib/job_tools/td:/studio/shrek3/shdw_td/lib/job_tools/surfacing:/studio/shrek3/shdw_td/lib/job_tools/sqhair_dev.args:/studio/shrek3/shdw_td/lib/job_tools/sqhair_dev:/studio/shrek3/shdw_td/lib/job_tools/run:/studio/shrek3/shdw_td/lib/job_tools/mpaint:/studio/shrek3/shdw_td/lib/job_tools/lighting:/studio/shrek3/shdw_td/lib/job_tools/layout:/studio/shrek3/shdw_td/lib/job_tools/fx:/studio/shrek3/shdw_td/lib/job_tools/extern:/studio/shrek3/shdw_td/lib/job_tools/char:/studio/shrek3/shdw_td/lib/job_tools/cfx:/studio/shrek3/shdw_td/lib/job_tools/anim:/studio/shrek3/shdw_td/lib/job_tools/anim:/studio/shrek3/shdw_td/lib/job_tools/td:/rel/map/shrek3-2005.11.3/bin:/rel/map/shrek3-cs/bin:/rel/map/percs-cs/bin:/usr/kerberos/bin:/usr/local/sys/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/opt/cxoffice/bin:/home/oracle/product/8.1.7/bin:/usr/java/bin:/rel/third_party/lsf/rwc/current/b
 in:/rel/third_party/lsf_share/rwc/current/scripts
    PERL5LIB=
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @pjcj

On Thu\, Dec 08\, 2005 at 11&#8203;:04&#8203;:16AM \-0800\, Jerry Hebert wrote&#8203;:

> Hello\, I have discovered an error in Data&#8203;::Dumper on versions
> greater than 5\.8\.0 Perl\. After 5\.8\.0\, Data&#8203;::Dumper uses qsortsv\(\)
> to sort keys if $Sortkeys is true and prior to that it used a
> pure Perl sort\. If you are dumping a hash\, this somehow causes
> problems with the hash iterator when attempting to subsequently
> use each\(\)\.
> 
> The following is a small script to illustrate the behavior\.
> 
> \#\!/usr/local/bin/perl
> use strict;
> use warnings;
> 
> use Data&#8203;::Dumper;
> 
> my $opts = \(  x => 5 \);

I think this should be my %opts \.\.\.

> $Data&#8203;::Dumper&#8203;::Sortkeys = 1;
> my $dumped = Data&#8203;::Dumper\->Dump \(\[ \\%opts \]\);
> 
> \# Uncomment the following keys\(\) line to print each properly
> \#keys %opts;
> while \(my \($key\, $val\) = each %opts\) \{
>     print "$key&#8203;: $val\\n";
> \}
> 
> \_\_END\_\_
> 
> You will see that the each\(\) loop prints nothing\. If you change 
> $Data&#8203;::Dumper&#8203;::Sortkeys to 0 or don't set it\, the 
> behavior will go away\. Also\, using $Data&#8203;::Dumper&#8203;::Useperl =1
> fixes the problem as well\.
> 
> Please let me know if you have further questions about this\.

This seems to have been fixed both in bleadperl and in maintperl which
is the codebase for release 5\.8\.8\.

\-\- 
Paul Johnson \- paul@&#8203;pjcj\.net
http&#8203;://www\.pjcj\.net

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ysth

On Thu\, Dec 08\, 2005 at 11&#8203;:04&#8203;:16AM \-0800\, Jerry Hebert wrote&#8203;:
> Hello\, I have discovered an error in Data&#8203;::Dumper on versions
> greater than 5\.8\.0 Perl\. After 5\.8\.0\, Data&#8203;::Dumper uses qsortsv\(\)
> to sort keys if $Sortkeys is true and prior to that it used a
> pure Perl sort\. If you are dumping a hash\, this somehow causes
> problems with the hash iterator when attempting to subsequently
> use each\(\)\.
> 
> The following is a small script to illustrate the behavior\.
> 
> \#\!/usr/local/bin/perl
> use strict;
> use warnings;
> 
> use Data&#8203;::Dumper;
> 
> my $opts = \(  x => 5 \);
> 
> $Data&#8203;::Dumper&#8203;::Sortkeys = 1;
> my $dumped = Data&#8203;::Dumper\->Dump \(\[ \\%opts \]\);
> 
> \# Uncomment the following keys\(\) line to print each properly
> \#keys %opts;
> while \(my \($key\, $val\) = each %opts\) \{
>     print "$key&#8203;: $val\\n";
> \}
> 
> \_\_END\_\_
> 
> You will see that the each\(\) loop prints nothing\. If you change 
> $Data&#8203;::Dumper&#8203;::Sortkeys to 0 or don't set it\, the 
> behavior will go away\. Also\, using $Data&#8203;::Dumper&#8203;::Useperl =1
> fixes the problem as well\.

This was fixed by \#25308 and will be in 5\.8\.8\.
As a workaround\, use a void\-context keys %hash just after dumping\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
